### DISCUSSION TITLE
brew bottle creates bottles with incorrect filenames

### DISCUSSION THREAD

**Discussion by ghost** [Homebrew/discussions#4541] [Tap maintenance and Homebrew development] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.19-24-ge96fd24
ORIGIN: https://github.com/Homebrew/brew
HEAD: e96fd24a7328a4913de2f7c56fd2c97d9aaa4572
Last commit: 5 days ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 2847fb559c8c8d3c2a7b41a05cab781929cd26a1
Core tap last commit: 22 hours ago
Core tap branch: master
Core tap JSON: 29 May 23:53 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.V42JvqHFZ7/org.xquartz:0
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 14.0.3 build 1403
Git: 2.40.1 => /opt/homebrew/bin/git
Curl: 7.88.1 => /usr/bin/curl
macOS: 13.4-arm64
CLT: 14.3.0.0.1.1679647830
Xcode: 14.3
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  platypus

Warning: Putting non-prefixed coreutils in your path can cause GMP builds to fail.

Warning: Unbrewed dylibs were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected dylibs:
  /usr/local/lib/libpas2jslib.dylib

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  dart
```

### Description of issue

While building bottles from homebrew formulae, I noticed the `brew bottle` is creating tarballs and json files with incorrect filenames. The tarballs should be named like `<name>-<version>.<arch>_<macos version>.bottle.<rebuild>.tar.gz`, but the generated tarballs have two dashes between `<name>` and `<version>`. This is making it impossible to install the newly-built bottle, because `brew install` looks for the wrong filename

How to reproduce:

1. Set up a local tap, and copy the `homebrew/core/hello` formula to the new tap
2. Install and bottle the hello formula

```
$ brew install --build-bottle charitybell/testing/hello
...
$ brew bottle --json --root-url http://localhost:8000 charitybell/testing/hello
==> Determining charitybell/testing/hello bottle rebuild...
==> Bottling hello--2.12_98.arm64_ventura.bottle.1.tar.gz...
./hello--2.12_98.arm64_ventura.bottle.1.tar.gz
  bottle do
    root_url "http://localhost:8000"
    rebuild 1
    sha256 cellar: :any, arm64_ventura: "adfc8d08174b2ae54a20eb2df6ae38d6f0b9b93fcc14390bf0a8d4cf8b5779ae"
  end
```

Notice that the bottle tar.gz filename is wrong. It should be `hello-2.12_98.arm64_ventura.bottle.1.tar.gz`, with only one dash after the formula name instead of two

3. Add the `bottle do` block to the charitybell/testing/hello formula
4. Start a webserver in the directory with the bottle
```
python -m http.server -b localhost
Serving HTTP on ::1 port 8000 (http://[::1]:8000/) ...
```

5. Try to install the bottle
```
brew install charitybell/testing/hello
==> Fetching charitybell/testing/hello
Error: hello: Failed to download resource "hello"
Failure while executing; `/usr/bin/env /opt/homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.0.19-24-ge96fd24\ \(Macintosh\;\ arm64\ Mac\ OS\ X\ 13.4\)\ curl/7.88.1 --header Accept-Language:\ en --retry 3 --fail --location --silent --head --request GET http://localhost:8000/hello-2.12_98.arm64_ventura.bottle.1.tar.gz` exited with 22. Here's the output:
curl: (22) The requested URL returned error: 404
HTTP/1.0 404 File not found
Server: SimpleHTTP/0.6 Python/3.11.3
Date: Mon, 29 May 2023 22:54:52 GMT
Connection: close
Content-Type: text/html;charset=utf-8
Content-Length: 335
```

The brew install command tried to download `hello-2.12_98.arm64_ventura.bottle.1.tar.gz`, but the only available bottle has the incorrect filename: `hello--2.12_98.arm64_ventura.bottle.1.tar.gz`

Here's the bottle's JSON file. The bottle `filename` is correct, and different from the bottle `local_filename`. I don't really know what to make of this. Is it a bug? Am I running one of the commands wrong? Should I just rename the file?

```json
{
  "charitybell/testing/hello": {
    "formula": {
      "name": "hello",
      "pkg_version": "2.12_98",
      "path": "Library/Taps/charitybell/homebrew-testing/Formula/hello.rb",
      "tap_git_path": "Formula/hello.rb",
      "tap_git_revision": "f50db20aba6f96ff79ebc58c0a646edb7a377b8e",
      "tap_git_remote": "https://github.com/charitybell/homebrew-testing",
      "desc": "gnu hello",
      "license": "AGPL-3.0-only",
      "homepage": "homepage"
    },
    "bottle": {
      "root_url": "http://localhost:8000",
      "cellar": "any",
      "rebuild": 1,
      "date": "2023-05-29",
      "tags": {
        "arm64_ventura": {
          "filename": "hello-2.12_98.arm64_ventura.bottle.1.tar.gz",
          "local_filename": "hello--2.12_98.arm64_ventura.bottle.1.tar.gz",
          "sha256": "3c5d249c1c813f5dddabe32f5a807aa1d34fb097b513340e103f80a09cf04e9b",
          "tab": {
            "homebrew_version": "4.0.19-24-ge96fd24",
            "changed_files": [
              "INSTALL_RECEIPT.json"
            ],
            "source_modified_time": 1643655444,
            "compiler": "clang",
            "runtime_dependencies": [
              {
                "full_name": "gettext",
                "version": "0.21.1",
                "declared_directly": false
              }
            ],
            "arch": "arm64",
            "built_on": {
              "os": "Macintosh",
              "os_version": "macOS 13",
              "cpu_family": "arm_firestorm_icestorm",
              "xcode": "14.3",
              "clt": "14.3.0.0.1.1679647830",
              "preferred_perl": "5.30"
            }
          }
        }
      }
    }
  }
}
```

**Comment by carlocab** [comment]
@Homebrew/brew, ideas?

**Comment by Bo98** [comment, marked as answer]
This is intentional. It's a weird quirk that someone can probably write the history of (it goes back to before I even was a maintainer), but backwards compatibility will appear somewhere in that story. It's difficult to change now without breaking existing workflows.

Note in the JSON you have:

```json
          "filename": "hello-2.12_98.arm64_ventura.bottle.1.tar.gz",
          "local_filename": "hello--2.12_98.arm64_ventura.bottle.1.tar.gz",
```

`local_filename` is the filename that `brew bottle` outputs while `filename` is the name expected on standard HTTP remotes. That is the reason why there's two entries in the JSON.

It is the responsibility of the uploader to select the correct remote filename on your server. For HTTP bottle servers, refer to what brew's GitHub Releases uploader does: https://github.com/Homebrew/brew/blob/33a3f72fbcf11cd9d0749198f768d68e274f2fa1/Library/Homebrew/github_releases.rb#L34-L39.

**Reply by ghost** [reply]
Great explanation, thank you! Good to know that it's intentional and that there's an easy solution :)

**Reply by atchox** [reply]
I was wondering what would be the way to resolve this if we are building and uploading bottles to the release using `softprops/gh-action-release`. Can we specify a different url (without the extra hyphen) against the filename?

_edit: nvm, renaming the files solves my issue_

**Comment by henrik242** [comment]
@Bo98  How about adding a flag to fix that?  ```brew bottle --correct-filename my-package```

**Reply by anttiharju** [reply]
Opened an issue, interested in seeing something like this be implemented
* https://github.com/Homebrew/brew/issues/19970
